build-tools/packages/build-cli/test/commands/typetests.test.ts

```diff
@@ -2,11 +2,162 @@
  * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
  * Licensed under the MIT License.
  */
-import { assert } from "chai";
+import { assert, expect } from "chai";
+import { PackageJson } from "@fluidframework/build-tools";
 
-import { previousVersion } from "../../src/commands/typetests";
+import {
+	VersionOptions,
+	applyTypeTestVersionOptions,
+	previousVersion,
+	resetBrokenTests,
+} from "../../src/commands/typetests";
+
+/**
+ * A minimal test package.json. It defines only the required fields according to the type definition.
+ */
+function packageMinimal(): PackageJson {
+	return {
+		name: "test-package",
+		version: "2.0.0-internal.6.0.0",
+		scripts: {},
+	};
+}
+
+/**
+ * A test package.json with an enabled typeValidation node.
+ */
+function packageWithTypeValidation(): PackageJson {
+	return {
+		...packageMinimal(),
+		devDependencies: {
+			"test-package-previous": "npm:test-package@2.0.0-internal.4.0.0",
+			"another-dependency": "^1.0.0",
+		},
+		typeValidation: {
+			broken: {
+				"broken-package": {
+					backCompat: false,
+					forwardCompat: false,
+				},
+			},
+			disabled: false,
+		},
+	};
+}
+
+/**
+ * A test package.json with a disabled typeValidation node.
+ */
+function packageWithTypeValidationDisabled(): PackageJson {
+	return {
+		...packageMinimal(),
+		devDependencies: {
+			"test-package-previous": "npm:test-package@2.0.0-internal.4.0.0",
+			"another-dependency": "^1.0.0",
+		},
+		typeValidation: {
+			broken: {
+				"broken-package": {
+					backCompat: false,
+					forwardCompat: false,
+				},
+			},
+			disabled: true,
+		},
+	};
+}
 
 describe("typetests tests", () => {
+	describe("applyTypeTestVersionOptions", () => {
+		describe("VersionOptions.Clear", () => {
+			it("removes previous test package dependency", () => {
+				const pkg = packageWithTypeValidation();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Clear);
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
```

Comment:
when asserting that soemthing doesn't exist, its really easy to have a bug in your test where it never exists instead of was properly removed. (Imagine the case where you had a typo in "test-package-previous", either here or in the origional data: the test would pass even if applyTypeTestVersionOptions didn't work)

If you add an assert that it does exist before you try and remove it (or make this super clear by having the input data in the test dorectly), its a bit more robust.

Another way to reduce risk of this issue is asserting the parent object is equal to something. That way you don't have to explicitly name the thing you expect to be missing so you can't typo it. Ex:
`assert.deepEqual(pkg.devDependencies, {})` (This would work if you didn't have the extra dep in this tests input)

Reply:
> If you add an assert that it does exist before you try and remove it (or make this super clear by having the input data in the test dorectly), its a bit more robust.

I really like the clarity this added. I _may_ have gone overboard, but in places where the input data is created elsewhere (like when using the factyory functions) checking before and after seems to add clarity and protect against test data changes that would silently pass.

> Another way to reduce risk of this issue is asserting the parent object is equal to something. That way you don't have to explicitly name the thing you expect to be missing so you can't typo it. Ex: assert.deepEqual(pkg.devDependencies, {}) (This would work if you didn't have the extra dep in this tests input)

I used this in a couple of places where it seemed clearer. But I didn't use it everywhere I could have because it seems misleading to assert the parent is empty. For example:

```ts
const pkg = packageWithTypeValidation(/* enabled */ false);
expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
updateTypeTestDependency(pkg, VersionOptions.ClearIfDisabled);
expect(pkg.devDependencies?.["test-package-previous"]).to.not.exist;
```

That seems clearer and more accurate to the intent of the test than:

```ts
const pkg = packageWithTypeValidation(/* enabled */ false);
expect(pkg.devDependencies?.["test-package-previous"]).to.exist;
updateTypeTestDependency(pkg, VersionOptions.ClearIfDisabled);
expect(pkg.devDependencies).to.deep.equal({});
```